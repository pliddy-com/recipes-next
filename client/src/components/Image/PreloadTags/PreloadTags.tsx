import Head from 'next/head';
import {
  AspectRatio,
  Breakpoints,
  calculateAspectRatio,
  createMediaQuery,
  createSrcSet
} from 'lib/responsiveImage';

interface PreloadTagsProps {
  aspectRatio?: AspectRatio;
  breakpoints: Breakpoints[];
  defaultWidth: number;
  url: string;
}

const PreloadTags = ({
  aspectRatio = '4 / 3',
  breakpoints,
  url,
  defaultWidth
}: PreloadTagsProps) => {
  const defaultHeight = Math.ceil(
    defaultWidth * calculateAspectRatio({ aspectRatio })
  );

  return breakpoints && breakpoints.length > 0 && url ? (
    <Head>
      {breakpoints.map(
        ({ viewMin, imgWidth }, index, breakpoints) =>
          url && (
            <link
              rel="preload"
              href={`${url}?w=${defaultWidth * 2}&h=${
                defaultHeight * 2
              }&fm=webp&q=75`}
              as="image"
              sizes={`${imgWidth}px`}
              imageSrcSet={createSrcSet({ url, imgWidth })}
              key={`${url}-${viewMin || imgWidth}-preload`}
              media={createMediaQuery({
                viewMin,
                index,
                breakpoints
              })}
              type="image/webp"
              fetchpriority="high"
            />
          )
      )}
    </Head>
  ) : null;
};

export default PreloadTags;
