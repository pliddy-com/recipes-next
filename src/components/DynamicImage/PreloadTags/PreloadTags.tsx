import Head from 'next/head';
import {
  createMediaQuery,
  createSrcSet,
  WidthProps,
} from 'lib/responsiveImage';

export interface PreloadTagsProps {
  props: WidthProps[];
  url: string;
}

// TODO: move PreloadTags to it's own component

export const PreloadTags = ({ props, url }: PreloadTagsProps) => (
  <Head>
    {props.map(({ viewMin, imgWidth }, index, propList) => (
      <link
        rel="preload"
        as="image"
        href={url || ''}
        imageSrcSet={createSrcSet({ url: url || '', imgWidth })}
        media={createMediaQuery({
          viewMin,
          index,
          propList,
        })}
        key={`${url}-${viewMin || imgWidth}-preload`}
        imageSizes={`${imgWidth}px`}
      />
    ))}
  </Head>
);

export default PreloadTags;
