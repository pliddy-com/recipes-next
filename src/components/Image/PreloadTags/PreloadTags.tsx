import Head from 'next/head';
import responsiveImage, { Breakpoints } from 'lib/responsiveImage';

interface PreloadTagsProps {
  breakpoints: Breakpoints[];
  url: string;
}

const PreloadTags = ({ breakpoints, url }: PreloadTagsProps) => {
  const { createSrcSet, createMediaQuery } = responsiveImage;

  return breakpoints && breakpoints.length > 0 && url ? (
    <Head>
      {breakpoints.map(
        ({ viewMin, imgWidth }, index, breakpoints) =>
          url && (
            <link
              rel="preload"
              href={url}
              as="image"
              imageSizes={`${imgWidth}px`}
              imageSrcSet={createSrcSet({ url, imgWidth })}
              key={`${url}-${viewMin || imgWidth}-preload`}
              media={createMediaQuery({
                viewMin,
                index,
                breakpoints,
              })}
            />
          )
      )}
    </Head>
  ) : null;
};

export default PreloadTags;
