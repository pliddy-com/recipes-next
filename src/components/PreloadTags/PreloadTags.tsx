import Head from 'next/head';
import responsiveImage, { WidthProps } from 'lib/responsiveImage';

interface PreloadTagsProps {
  props: WidthProps[];
  url: string;
}

const PreloadTags = ({ props, url }: PreloadTagsProps) => {
  const { createSrcSet, createMediaQuery } = responsiveImage;

  return props && props.length > 0 && url ? (
    <Head>
      {props.map(
        ({ viewMin, imgWidth }, index, propList) =>
          url && (
            <link
              as="image"
              href={url}
              imageSizes={`${imgWidth}px`}
              imageSrcSet={createSrcSet({ url, imgWidth })}
              key={`${url}-${viewMin || imgWidth}-preload`}
              media={createMediaQuery({
                viewMin,
                index,
                propList,
              })}
              rel="preload"
            />
          )
      )}
    </Head>
  ) : null;
};

export default PreloadTags;
