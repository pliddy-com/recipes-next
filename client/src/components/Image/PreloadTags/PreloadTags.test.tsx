import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import PreloadTags from './PreloadTags';

jest.mock('next/head');

describe('PreloadTags', () => {
  afterEach(() => {
    jest.resetModules();
  });

  describe('when props and url are provided', () => {
    const url = 'https://URL.test';
    const breakpoints = [
      {
        viewMin: 1000,
        imgWidth: 300
      },
      {
        viewMin: 500,
        imgWidth: 200
      },
      {
        imgWidth: 100
      }
    ];

    const defaultWidth = breakpoints[0].imgWidth;

    it('it renders a set of preload link tags in a Head component', () => {
      const expectedImageSizes = '300px';

      const { asFragment } = render(
        <PreloadTags
          breakpoints={breakpoints}
          defaultWidth={defaultWidth}
          url={url}
        />
      );

      // returns 1 tag for each element in props
      const linkTags = document.getElementsByTagName('link');
      expect(linkTags.length === breakpoints.length).toBe(true);

      // expect url & imageSizes tags to be created
      const testTag = linkTags[0];
      expect(testTag.textContent?.includes(url));
      expect(testTag.textContent?.includes(expectedImageSizes));

      // assert that the component matches the existing snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when props and url are provided', () => {
    it('it does not render', () => {
      render(
        <PreloadTags
          aspectRatio="4 / 3"
          breakpoints={[]}
          defaultWidth={0}
          url={null as unknown as string}
        />
      );

      const linkTags = document.getElementsByTagName('link');

      expect(linkTags.length).toBe(0);
    });
  });
});
