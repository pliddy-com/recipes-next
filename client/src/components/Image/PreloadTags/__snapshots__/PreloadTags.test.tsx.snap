// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreloadTags when props and url are provided it renders a set of preload link tags in a Head component 1`] = `
<DocumentFragment>
  <link
    as="image"
    fetchpriority="high"
    href="https://URL.test?w=600&fm=webp&q=75"
    imagesrcset="https://URL.test?w=300&fm=webp&q=75 1x, https://URL.test?w=600&fm=webp&q=75 2x"
    media="(min-width: 1000px)"
    rel="preload"
    sizes="300px"
    type="image/webp"
  />
  <link
    as="image"
    fetchpriority="high"
    href="https://URL.test?w=600&fm=webp&q=75"
    imagesrcset="https://URL.test?w=200&fm=webp&q=75 1x, https://URL.test?w=400&fm=webp&q=75 2x"
    media="(min-width: 500px) and (max-width: 999px)"
    rel="preload"
    sizes="200px"
    type="image/webp"
  />
  <link
    as="image"
    fetchpriority="high"
    href="https://URL.test?w=600&fm=webp&q=75"
    imagesrcset="https://URL.test?w=100&fm=webp&q=75 1x, https://URL.test?w=200&fm=webp&q=75 2x"
    media="(max-width: 499px)"
    rel="preload"
    sizes="100px"
    type="image/webp"
  />
</DocumentFragment>
`;
